### ISSUE TITLE
Any way to toggle show/hide hidden directories and files?

### ISSUE THREAD

**Issue by pfr-dev** [LeperGnome/bt#8]
Firstly, let me just say thank you. This program is awesome and I will be switching to it permanently once it's close to feature complete. I've used `nnn`, `fff`, `shfm`, `ranger`, `lf` etc. for a long time. But this implementation using the `tree` is just awesome!

I wonder if it could be possible to toggle show/hide hidden dirs and files? that would be neat. I assume that if you're leveraging the `tree` command then simply dropping the `-a` flag should do it?

I'm also excited for you're implementation of image previews.. I'm curious what route you might take for this?

**EDIT**
I'm also curious how to set default programs for opening files? for example, I'm attempting to open an image file but it obviously doesn't know with which program to open it. where can I set this? Is there an `opener` file that it needs to read?

**Comment by LeperGnome** [comment]
Hi! Thanks a lot, I'm really glad, that you like this project! 
First of all, no, I'm not using `tree`, all the rendering is custom, so I would not be able to just drop `-a`, but regardless, toggling hidden files view should not be too hard to implement.

About image previews - yea, I'm also excited about this! I wasn't giving it too much yet though, but currently I think that I'll choose different strategies depending on terminal emulator with a fallback of covering image to a set of colorful half-blocks. You can share your suggestions! 

About opening files - currently `bt` only opens files in `$EDITOR` on `e`, because it's a primary use-case... for me =) I'll definitely implement opening files with different programs (leveraging `open`? Not yet sure)

**Comment by pfr-dev** [comment]
Yeah these days you've got to cover all bases when it comes to images in the terminal. Kitty image protocol, sixel, ueberzug++, chafa etc... is a bit of a task I know.

However, implementing the use of an opener script first would say least allow the user to open image files with an image viewer. 

I use my terminal for browser to open all files, PDFs, XMLs, PNGs.. anything. So an opener script would be great. using XDG_OPEN would also work, but I l personally prefer to have fine grained control using a script. 

I currently use the following script with `shfm` and `fff`

```
#!/bin/sh -e
#
# open file in application based on file extension

case $(file -bi "$1") in
    image/png*|image/jpg*|image/jpeg*)
	feh -g 700x393 --scale-down "$1"
    ;;
    text/*)
    "${EDITOR:=vim}" "$1"
    ;;
    audio/*)
	mpv --no-video "$1"
    ;;
    video/*)
	mpv --geometry=700x393 "$1"
    ;;
    *.html)
    firefox "$1"
    ;;
    *.pdf)
	zathura "$1"
    ;;
    *) # all other file types
    xdg-open "$1"
    ;;
esac
```

**Comment by LeperGnome** [comment]
Noted, thank you! Customizable openers for different filetypes look great, I would love to implement it!

**Comment by LeperGnome** [comment]
@pfr-dev Hey! I sat down to think about your feature request and made a quick PR, but I'm kinda hesitant... I'll be glad if you cared to look at my approach and let me know what you think. Do you think, that toggling is fine (or even better) only for current directory and not the whole tree?

**Comment by LeperGnome** [comment]
Work kinda like this. As you can see, I can toggle hidden files only on a single "depth level".
[Screencast from 18.11.2024 02:36:03.webm](https://github.com/user-attachments/assets/3e5ea012-956e-4d2a-a4ec-6d846c4091f8)

**Comment by pfr-dev** [comment]
Hmmmm cool idea! But... I think if you want to only toggle hidden files within a current directory, then it should also include all subdirectories. Otherwise, it might get confusing if you have to keep toggling fore every subdirectory. 

What to you think?

**Comment by LeperGnome** [comment]
I'm not sure here. What if you want to disable hidden for current, and enable for children? You will just need to toggle it in every child. So I think, that it's more or less the same in terms of functionality, could be a bit more confusing from my perspective, but if I'm being honest, I don't necessarily like either option xD

I looked at how it's done in `ranger`, and, well, it's just easier to implement, since there's max of 3 opened directories shown on the screen, so they can just refresh all 3.

**Comment by LeperGnome** [comment]
May be I could make some sort of indicator, that will show, that hidden directories are actually hidden... Idk, need to experiment

**Comment by LeperGnome** [comment]
The small dot near the directory name to indicate, that hidden files are not shown at this level. Exciting? Not for me really, but might get the job done. (p.s. hated all the emojies and larger signs) 
![image](https://github.com/user-attachments/assets/63607a54-5405-4db9-a057-dfcf1d9c71cc)

**Comment by pfr-dev** [comment]
I guess here-in lies the problem with a tree style file manager. If you've got Lots and lots of files, it quickly becomes difficult to navigate. But I see what you're saying, and I agree, an indicator to let you know that a directory contains hidden files is a good idea and you're example above looks perfect!

I have one more suggestion which might help with space saving while navigating large directories. 

Please let me know if you'd like me to split this into another issue, but for now I feel it's kind of relevant here.

What if, when you navigate to a subdirectory, the parent directory collapses, only showing the name of the parent directory but not every other subdirectory there within?

**Comment by LeperGnome** [comment]
Well, It's an interesting idea, but if I understood you correctly, it will just remove all the other context, that could be helpful. E.g. I want to move `./pkg/stack/stack.go` to `./cmd/bt/main.go`. Then, instead of seeing this, and being able to plan all my movements
```
.
├─ .git
├─ assets
├─ bin
├─ cmd
│  └─ bt
│     └─ main.go
├─ internal
├─ pkg
│  └─ stack <-
│     └─ stack.go
├─ test
├─ .gitignore
├─ go.mod
├─ go.sum
├─ LICENSE.txt
├─ Makefile
└─ README.md
```
I'll see this: 
```
.
├─ pkg
│  └─ stack <-
│     └─ stack.go
```
No real point of having a tree structure with such rendering. 
Please correct me if I misunderstood you

**Comment by LeperGnome** [comment]
I think, that if space is the issue, then it may be should be solved by other means, like faster up/down movement (like Ctrl+u/d in vim), or searching filename, or leaving marks.
I honestly just really like how file trees look, It's a very intuitive representation and I don't want to give it up =)

**Comment by pfr-dev** [comment]
Yeah ok, you're right (again!) haha I think this works well. The indicator idea is great because I'll generally know which subdirectory my needed dotfiles are. 

Has this been implemented in the master tree yet? Would live to test it out.

**Comment by LeperGnome** [comment]
> Has this been implemented in the master tree yet? Would love to test it out.

nope, just in pr yet 

Thanks man, I really appreciate your interest in this project! 
I actually initially wrote a similar program in rust - [trre](https://github.com/LeperGnome/trre), but recursive data structures are just a paint to work with there (or it's my skill issue, most likely), so I switched to go, GC just makes life easier =)

**Comment by pfr-dev** [comment]
Back on the custom `opener` script suggester earlier.. Why not just have a `bt.conf` config file where you can set default programs for opening files as well as being able to customize the keybindings? I'm one of those weirdos who likes to swap `j` and `k` in my `.vimrc` so I'd love to have the ability to tweak the keybindings also.

**Comment by LeperGnome** [comment]
Yea, I would live to add more customization!

**Comment by LeperGnome** [comment]
I've been inactive for some time, but came here to merge my MR #9 =)

**Comment by pfr-dev** [comment]
Doesn't appear to be merged into the main branch yet?

I've had to clone the git repo and install using 
```
$ git checkout feat/toggle-hidden
$ go install ./...
```

**Comment by LeperGnome** [comment]
@pfr-dev Hey! Feature should be in the latest release, along with some other thing. Check it out, if you haven't already!
